### PR TITLE
Fixing n-i-t launcher

### PR DESCRIPTION
**Pull request by yschimke**
Graal main method to run tests with minimal reflection and automatic settings.
Uses the test list in native-image-tests/src/main/resources/testlist.txt.

### CODE REVIEW

**Comment by yschimke** [comment]
@swankjesse With RunTests.kt ultimately with enough cleanup we may get to a point that ConsoleLauncher would be cleanly usable instead of this classlist.txt approach, but I've found that reflection and loading all available classes opportunistically with GraalVM is very brittle.  I'd like this to be both simple (no extra prestep like manual run of seed) and predictable (doesn't trigger a lot of interesting code paths accidentally).

File: native-image-tests/build.gradle [hunk 1]
```diff
@@ -20,7 +20,7 @@ dependencies {
   implementation project(':mockwebserver-deprecated')
   implementation project(':okhttp-urlconnection')
   implementation project(':mockwebserver-junit4')
-  testRuntimeOnly project(':mockwebserver-junit5')
+  implementation project(':mockwebserver-junit5')
```
Comment (swankjesse) [inline native-image-tests/build.gradle:23]: This one’s interesting. We don’t refer to it at compile time right?

File: settings.gradle [hunk 2]
```diff
@@ -11,10 +11,7 @@ if (properties.containsKey('android.injected.invoked.from.ide') ||
   include ':android-test'
 }
 
-if (!properties.containsKey('android.injected.invoked.from.ide')) {
```
Comment (swankjesse) [inline settings.gradle:14]: Nice improvement here

File: native-image-tests/src/main/kotlin/okhttp3/TestRegistration.kt [hunk 3]
```diff
@@ -18,25 +18,43 @@ package okhttp3
...
+    knownTestFile.readLines().forEach {
+      try {
+        val testClass = access.findClassByName(it)
+
+        if (testClass != null) {
+          access.registerAsUsed(testClass)
+          registerTest(testClass)
+        }
+      } catch (e: Exception) {
+        e.printStackTrace()
```
Comment (swankjesse) [inline native-image-tests/src/main/kotlin/okhttp3/TestRegistration.kt:53]: rethrow? what’s this defending against

File: native-image-tests/src/main/kotlin/okhttp3/RunTests.kt [hunk 4]
```diff
@@ -17,60 +17,87 @@ package okhttp3
...
+    .lines()
+    .filter { it.isNotBlank() }
+    .map {
+      selectClass(Class.forName(it, false, sampleTestClass.classLoader))
+    }
+}
+
+fun buildRequest(selectors: List<DiscoverySelector>): LauncherDiscoveryRequest {
+  val request: LauncherDiscoveryRequest = LauncherDiscoveryRequestBuilder.request()
+    // TODO replace junit.jupiter.extensions.autodetection.enabled with API approach.
```
Comment (swankjesse) [inline native-image-tests/src/main/kotlin/okhttp3/RunTests.kt:84]: what’s this mean? Changing how we configure JUnit here? Or how we did it in our tests?

File: native-image-tests/src/main/kotlin/okhttp3/GenerateClassList.kt [hunk 5]
```diff
@@ -0,0 +1,15 @@
+package okhttp3
+
+import org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor
+import java.io.File
+
+/**
+ * Run periodically to refresh the known set of working tests.
+ *
+ * TODO use filtering to allow skipping acceptable problem tests
```
Comment (swankjesse) [inline native-image-tests/src/main/kotlin/okhttp3/GenerateClassList.kt:9]: Better to integrate with PlatformTest, like Platform.isGraalvm() ?

File: native-image-tests/src/main/kotlin/okhttp3/GenerateClassList.kt [hunk 6]
```diff
@@ -0,0 +1,15 @@
+package okhttp3
```
Comment (swankjesse) [inline native-image-tests/src/main/kotlin/okhttp3/GenerateClassList.kt:1]: supernit: copyright

**Review by swankjesse — APPROVED** [review]

File: native-image-tests/build.gradle (hunk 1 above)
Comment (yschimke) [reply native-image-tests/build.gradle:23]: There is no test runtime!

File: native-image-tests/src/main/kotlin/okhttp3/RunTests.kt (hunk 4 above)
Comment (yschimke) [reply native-image-tests/src/main/kotlin/okhttp3/RunTests.kt:84]: https://github.com/junit-team/junit5/issues/2470

I'm this close to a config driven completely without any magic.  But then I nee to set a system property to make it use JAR services mechanism.  I'd prefer someway to do that in code.

File: native-image-tests/src/main/kotlin/okhttp3/TestRegistration.kt (hunk 3 above)
Comment (yschimke) [reply native-image-tests/src/main/kotlin/okhttp3/TestRegistration.kt:53]: I'll comment. If you throw an exception here then native image building fails half way through siltently without rewriting the binary.  So we report noisily, but keep going and prefer running most tests still.

File: native-image-tests/src/main/kotlin/okhttp3/GenerateClassList.kt (hunk 5 above)
Comment (yschimke) [reply native-image-tests/src/main/kotlin/okhttp3/GenerateClassList.kt:9]: Yep, probably in a follow up. I'd still like to use tags. Go full junit5 native on this.

File: native-image-tests/src/main/kotlin/okhttp3/GenerateClassList.kt (hunk 6 above)
Comment (yschimke) [reply native-image-tests/src/main/kotlin/okhttp3/GenerateClassList.kt:1]: You should automate this

```
if pr.author == yschimke and files.new > 0
  comment("remember copyright")
```

**Pull request merged**